pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+}
+
+var _ directbase.Model = &entitlementModel{}
+
+type entitlementModel struct {
+	config config.ControllerConfig
+}
+
+func (m *entitlementModel) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building PrivilegedAccessManager client for Entitlement: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *entitlementModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+
+	obj := &krm.PrivilegedAccessManagerEntitlement{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	container, err := oneOfContainer(ctx, reader, obj,
+		obj.Spec.ProjectRef,
+		obj.Spec.FolderRef,
+		obj.Spec.OrganizationRef)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving 'obj.Spec.ProjectRef', "+
+			"'obj.Spec.FolderRef' and 'obj.Spec.OrganizationRef': %w", err)
+	}
+
+	// Get location
+	location := *obj.Spec.Location
+
+	var id *PrivilegedAccessManagerEntitlementIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(container, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.Container != container {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has parent container changed, expected %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Container, container)
+		}
+		if id.Parent.Location != location {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Location, location)
+		}
+		if id.Entitlement != resourceID {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Entitlement, resourceID)
+		}
+	}
+
+	container, err = oneOfContainer(ctx, reader, obj,
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.ProjectRef,
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.FolderRef,
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.OrganizationRef)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving 'obj.Spec.PrivilegedAccess.GcpIAMAccess.ProjectRef', "+
+			"'obj.Spec.PrivilegedAccess.GcpIAMAccess.FolderRef' and "+
+			"'obj.Spec.PrivilegedAccess.GcpIAMAccess.OrganizationRef': %w", err)
+	}
+	switch *obj.Spec.PrivilegedAccess.GcpIAMAccess.ResourceType {
```

Comment:
Ah I was wondering about this ... it might be nice to add a comment to ResourceType (not a blocker).  We could also consider making it optional, it is implied by projectRef / folderRef / organizationRef I think

Reply:
I plan to make this part of the improvement after we discuss the schema with the service team.

Personally, I find both (all) `resourceType` and `projectRef`/`folderRef`/`organizationRef` fields useless because they must match the parent project/folder/organization. We might want to hide all of them.